src/command_modules/azure-cli-storage/azure/cli/command_modules/storage/blob.py

```diff
@@ -124,3 +124,143 @@ def _copy_single_blob(source_blob):
 
     return [BlobCopyResult(b.name, _copy_single_blob(b).id) for b in source_blobs]
 
+
+#pylint: disable=unused-argument
+def storage_blob_download_batch(client, source, destination, source_container_name, pattern=None,
+                                progress=False):
+    """
+    Download blobs in a container recursively
+
+    :param str source:
+        The string represents the source of this download operation. The source can be the
+        container URL or the container name. When the source is the container URL, the storage
+        account name will parsed from the URL.
+
+    :param str destination:
+        The string represents the destination folder of this download operation. The folder must
+        exist.
+
+    :param str pattern:
+        The pattern is used for files globing. The supported patterns are '*', '?', '[seq]',
+        and '[!seq]'.
+
+    :param bool progress:
+        Shows the download progress.
+    """
+    import os.path
+    from fnmatch import fnmatch
+
+    def _pattern_has_wildcards(p):
+        return not p or p.find('*') != -1 or p.find('?') != -1 or p.find('[') != -1
+
+    source_blobs = []
+    if not _pattern_has_wildcards(pattern):
+        source_blobs.append(pattern)
+    else:
+        # IDEA:
+        # 1. Listing is slow. It can be done in parallel with copying.
+        # 2. Use the prefix parameter to reduce the returned blobs list size
+        source_blobs = client.list_blobs(source_container_name)
+
+        if pattern:
+            source_blobs = [blob.name for blob in source_blobs if fnmatch(blob.name, pattern)]
```

Comment:
to accumulate results and separate the logic of selecting and downloading. because I was also thinking of doing those two in parallel. the separation will make the work easiler.